### ISSUE TITLE
Optionally type check unannotated functions

### ISSUE THREAD

**Issue by JukkaL** [JukkaL/mypy#817]
Add a mode that runs the type checker for unannotated functions as well. Variables will get implicit `Any` types (except maybe for things where we can be certain of the type, e.g. if `x = []` is the only assignment to a local variable, we can infer type `List[Any]`).

Examples of things that could be caught:

* Access to undefined attributes in modules (e.g. `modname.misspelled`).
* Calls to functions with wrong argument counts, invalid keyword arguments or, in some cases, invalid argument types.
* Invalid operations involving primitive types (e.g., `'foo' + 1`).

`self` will probably still have type `Any`. Alternatively, we could check that calls to methods that are present in the class definition via `self` (and that don't have decorators) are okay. We probably must allow undefined attributes to be accessed as derived classes could define them to avoid false positives. Examples:

```
class A:
    def f(self):
        self.g(1)  # Report error, wrong number of arguments
        self.unknown()  # No error

    def g(self, x, y): ...
```

**Comment by o11c** [comment]
I'm thinking about adding a `-Wfoo` system similar to gcc for this sort of thing.

**Comment by kirbyfan64** [comment]
If this is going to be in, a compliment to the option should be added that excludes certain modules.

**Comment by JukkaL** [comment]
Yeah, something like that would be useful. We could also not use the mode for the Python std lib and 3rd party libraries -- i.e., only use aggressive checking for the target program and things explicitly added to `MYPYPATH` (note that currently we don't look into `PYTHONPATH` by default).

Alternatively, this mode could be opt-in on a file by file basis. Maybe something like this:

```
# mypy: check_all

def foo():
    1 + 'foo'   # Will be reported as an error because of the check_all flag
```

Or add a module-level special attribute for this purpose.
